src/v/cloud_roles/azure_aks_refresh_impl.cc

```diff
@@ -0,0 +1,195 @@
+#include <rapidjson/error/en.h>
+
+namespace {
+
+// utility to wrap std::getenv for required environment variables
+auto load_from_env(const char* env_var) -> ss::sstring {
+    auto env_value = std::getenv(env_var);
+    if (!env_value) {
+        throw std::runtime_error(fmt::format(
+          "environment variable {} is not set, the Azure AKS client cannot "
+          "function "
+          "without this.",
+          env_var));
+    }
+    return env_value;
+}
+
+constexpr static auto env_var_azure_client_id = "AZURE_CLIENT_ID";
+constexpr static auto env_var_azure_tenant_id = "AZURE_TENANT_ID";
+constexpr static auto env_var_azure_federated_token_file
+  = "AZURE_FEDERATED_TOKEN_FILE";
+constexpr static auto env_var_azure_authority_host = "AZURE_AUTHORITY_HOST";
+
+} // namespace
+
+namespace cloud_roles {
+
+azure_aks_refresh_impl::azure_aks_refresh_impl(
+  net::unresolved_address address,
+  aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    [&] {
+        if (!address.host().empty()) {
+            // non-empty host: it's an override that we should use
+            return std::move(address);
+        }
+        // empty host: use env_var
+        return net::unresolved_address{
+          load_from_env(env_var_azure_authority_host), default_port};
+    }(),
+    std::move(region),
+    as,
+    retry_params)
+  , client_id_{load_from_env(env_var_azure_client_id)}
+  , tenant_id_{load_from_env(env_var_azure_tenant_id)}
+  , federated_token_file_{load_from_env(env_var_azure_federated_token_file)} {}
+
+auto azure_aks_refresh_impl::print(std::ostream& os) const -> std::ostream& {
+    fmt::print(os, "azure_aks_refresh_impl{{address:{}}}", address());
+    return os;
+}
+
+auto azure_aks_refresh_impl::fetch_credentials() -> ss::future<api_response> {
+    // https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#third-case-access-token-request-with-a-federated-credential
+    // load the encoded jwt from the token file (its content is rotated by k8s),
+    // compose:
+    // POST /{tenant_id_}/oauth2/v2.0/token HTTP/1.1               // Line
+    // breaks for clarity Host: {authority_host_}:443 Content-Type:
+    // application/x-www-form-urlencoded
+    //
+    // scope=https%3A%2F%2Fgraph.microsoft.com%2F.default
+    // &client_id={client_id_}
+    // &client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer
+    // &client_assertion={content(federated_token_file_)}
+    // &grant_type=client_credentials
+
+    auto jwt_token_fut = co_await ss::coroutine::as_future(
+      read_fully_to_string({federated_token_file_}));
+    if (jwt_token_fut.failed()) {
+        vlog(
+          clrl_log.error,
+          "failed to read OIDC pod token from file {}",
+          federated_token_file_);
+        co_return ss::coroutine::exception(jwt_token_fut.get_exception());
+    }
```

Comment:
not something we should change right now, but it would be nice if there was a `start` or `init` or similar method in the impl  to do these one time initialization operations up front since we can't do these in the c-tor.

Reply:
not sure this is the case, but i think that jwt token can be rotated so it has to be read every time

Reply:
> not sure this is the case, but i think that jwt token can be rotated so it has to be read every time

this makes sense